### ISSUE TITLE
Use model on Android?

### ISSUE THREAD

**Issue by pavviaz** [lukas-blecher/LaTeX-OCR#6]
Hi! Your model is working great on PC, but is is possible to use it on Android device?
As far as I know, the model have to be converted to TorchScript format to work on mobile device, but it's not enough. We also need to transfer "call_model" function from pix2tex.py script to Android app, because model requires specific image resize to work. How we can do that? Thank you :)

**Comment by lukas-blecher** [comment]
Hi, sorry for the delayed response.
I tried again to trace the model and still didn't get it to work although I did come closer this time. I think it is a bit problematic that the model accepts dynamic input sizes but definitely necessary especially if you want to run it on moblie.

As for the preprocessing in the `call_model` function, here is what it does
- Convert the image to grayscale
- crop it in a way that the image content hugs the left and upper image sided
- pad it to the right and bottom with white pixels until both image dimensions are divisible by 32 (this all happens in the [`pad`](https://github.com/lukas-blecher/LaTeX-OCR/blob/1b921750e2cba690ab29ce56e46a6276b6dc0db8/utils/utils.py#L73) function
- optionally resize the image to represent the training data more closely. Here another comes into play. (Not needed if the images are fairly small)
- normalize the pixel values as specified in the [`test_transform`](https://github.com/lukas-blecher/LaTeX-OCR/blob/1b921750e2cba690ab29ce56e46a6276b6dc0db8/dataset/dataset.py#L41)

That's basically it.
Glad to hear that you want to use the model. If you want to use the camera to take images it definitely need some more finetuning. Not only in the preprocessing step but also for the model. As of now the model can only work with screen snippets but I'm sure with the right dataset/augmentation it could work with other images as well.

**Comment by pavviaz** [comment]
So, after some tryhard, I finally managed to use your model in Android application, but the model itself is placed on the server, and the app only sends an image to server as a POST query, and the server giving back decoded equation. I tried to lauch the model on Android itself with Chaquopy framework and Python-for-Android, but it turned out that PyTorch in it's row view is impossible to launch on the mobile device, because it's kernel is written in c/c++

Actually, I am not really big professional in Neural Networks (begginer would be better word), and I've only just learned the working mechanism of such Neural Networks like im2latex (visual attention is the real thing), and I found great tutorials about that on TensorFlow. May be, it would be possible to create similar NN with TF and launch it directly on mobile device with TFlite?... Should give that a try :)

P.S.: Huge thanks for your repo! It's really helped me a lot in my University Android project
